### PR TITLE
pkg/{store,util}: add safeguards against panics in Series()

### PR DESCRIPTION
**Pull request by GiedriusS**
Add a new package which will help us safeguard against accidental panics
in the most important Series() call in Thanos Store. Use it in
pkg/store/bucket.go. We already have a top-level handler for panics but
it doesn't work for spawned goroutines so use this one for extra safety.

TODO:
* hook into the panics metric
* guard more places
* tests

### CODE REVIEW

File: pkg/util/util.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
+// Package util houses various utilities that do not fit anywhere else.
```
Comment (bwplotka) [inline pkg/util/util.go]: We should focus on removing those (: 

`util` is ALWAYS bad idea as it ends up being pile of trash for everything we are too lazy to think package name ;p

I think this should be actually part of `run`. We can try contribute to run.Group (can start github issue) or fork our own version, as this is must have.

File: pkg/util/util.go [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+	"github.com/pkg/errors"
+)
+
+// RecoverGoroutine is a wrapper to be used for calling new goroutines / using run.Group
+// which handles panics gracefully - prints the info out to the error log
+// and returns an error.
+func RecoverGoroutine(logger log.Logger, f func() error, extrafields ...interface{}) func() error {
+	return func() (err error) {
+		defer func() {
+			if p := recover(); p != nil {
```
Comment (bwplotka) [inline pkg/util/util.go]: We should decide how to handle errors.. Preferably just pass `func() error` to have caller decide about this?

**Comment by GiedriusS** [comment]
Unfortunately, `run.Group` will not be expanded to include this kind of option so we will have to do it ourselves 😞

**Comment by bwplotka** [comment]
Worth a run group fork/copy then? :thinking:

**Comment by GiedriusS** [comment]
@bwplotka do you think we should do it as part of our code base or move it into another repo? (:

**Comment by bwplotka** [comment]
Whatever works better for us. Happy to start it as Thanos package (:

**Comment by povilasv** [comment]
Maybe we can use https://github.com/grpc-ecosystem/go-grpc-middleware/tree/master/recovery for this instead of making our own package?

Usage example:
```
myServer := grpc.NewServer(
    grpc.StreamInterceptor(grpc_middleware.ChainStreamServer(
        ...
        grpc_recovery.StreamServerInterceptor(),
    )),
    grpc.UnaryInterceptor(grpc_middleware.ChainUnaryServer(
       ...
        grpc_recovery.UnaryServerInterceptor(),
    )),
)
```

**Comment by bwplotka** [comment]
We want to safeguard for panics inside the `rungroup/errgroup` goroutine.
This will be not caught by recovery in gRPC request main goroutine (:

Also we already use `grpc_recovery`

**Pull request closed without merging**